process/block/poolsCleaner/crossTxsPoolsCleaner.go

```diff
@@ -0,0 +1,306 @@
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedUnsignedTx", "hash", key)
+	ctpc.processReceivedTx(key, unsignedTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) processReceivedTx(key []byte, txType int8) {
+	isCrossTx := false
+	var crossTxInfo *txInfo
+	for shardID := uint32(0); shardID < ctpc.shardCoordinator.NumberOfShards(); shardID++ {
+		isCrossTx, crossTxInfo = ctpc.isCrossTxFromShard(key, shardID, txType)
+		if isCrossTx {
+			break
+		}
+	}
+
+	if !isCrossTx {
+		isCrossTx, crossTxInfo = ctpc.isCrossTxFromShard(key, core.MetachainShardId, txType)
+		if !isCrossTx {
+			return
+		}
+	}
+
+	if crossTxInfo == nil {
+		return
+	}
+
+	ctpc.mutMapCrossTxsRounds.Lock()
+	defer ctpc.mutMapCrossTxsRounds.Unlock()
+
+	if _, ok := ctpc.mapCrossTxsRounds[string(key)]; !ok {
+		crossTxInfo.round = ctpc.rounder.Index()
+		crossTxInfo.txType = txType
+		ctpc.mapCrossTxsRounds[string(key)] = crossTxInfo
+
+		log.Trace("transaction has been added",
+			"hash", key,
+			"round", crossTxInfo.round,
+			"sender shard", crossTxInfo.senderShardID,
+			"receiver shard", crossTxInfo.receiverShardID,
+			"type", getTxTypeName(crossTxInfo.txType))
+	}
+}
+
+func (ctpc *crossTxsPoolsCleaner) isCrossTxFromShard(txHash []byte, shardID uint32, txType int8) (bool, *txInfo) {
+	transactionPool := ctpc.getTransactionPool(txType)
+	if transactionPool == nil {
+		return false, nil
+	}
+
+	selfShardID := ctpc.shardCoordinator.SelfId()
+	if shardID == selfShardID {
+		return false, nil
+	}
+
+	strCache := process.ShardCacherIdentifier(shardID, selfShardID)
+	txStore := transactionPool.ShardDataStore(strCache)
+	if txStore == nil {
+		return false, nil
+	}
+
+	_, ok := txStore.Peek(txHash)
+	if !ok {
+		return false, nil
+	}
+
+	return true, &txInfo{txStore: txStore, senderShardID: shardID, receiverShardID: selfShardID}
+}
+
+func (ctpc *crossTxsPoolsCleaner) cleanCrossTxsPoolsIfNeeded() int {
+	ctpc.mutMapCrossTxsRounds.Lock()
+	defer ctpc.mutMapCrossTxsRounds.Unlock()
+
+	selfShardID := ctpc.shardCoordinator.SelfId()
+	numPendingMiniBlocks := ctpc.blockTracker.GetNumPendingMiniBlocks(selfShardID)
+	numTxsCleaned := 0
+
+	for hash, crossTxInfo := range ctpc.mapCrossTxsRounds {
+		value, ok := crossTxInfo.txStore.Get([]byte(hash))
+		if !ok {
+			log.Trace("transaction not found in pool",
+				"hash", []byte(hash),
+				"round", crossTxInfo.round,
+				"sender shard", crossTxInfo.senderShardID,
+				"receiver shard", crossTxInfo.receiverShardID,
+				"type", getTxTypeName(crossTxInfo.txType))
+			delete(ctpc.mapCrossTxsRounds, hash)
+			continue
+		}
+
+		txHandler, ok := value.(data.TransactionHandler)
+		if !ok {
+			log.Debug("cleanCrossTxsPoolsIfNeeded", "error", process.ErrWrongTypeAssertion,
+				"hash", []byte(hash),
+				"round", crossTxInfo.round,
+				"sender shard", crossTxInfo.senderShardID,
+				"receiver shard", crossTxInfo.receiverShardID,
+				"type", getTxTypeName(crossTxInfo.txType))
```

Comment:
you don't need to delete this tx from pool/local crossTxInfo?

Reply:
Actually if this happens, wrong type assertion error, probably there is a kind of garbage in pool. But yes, I would add these both removal here.